### PR TITLE
Training experiments

### PR DESCRIPTION
**Pull request by MatteoKartoon**
- `lr_decay_rate`, `lr_decay_epochs` and `fine_tune_last` set in the bash script
- gradient clipping
- saving predictions over green background implemented in `visualization.py`
- testset average implemented in `evaluation.py`
- small bug fixed
- switch to `briaai1.4`

### CODE REVIEW

File: birefnet/config.py [hunk 1]
```diff
@@ -8,13 +8,13 @@
 DEFAULT_BCE_WITH_LOGITS = False
 DEFAULT_LAMBDAS_PIX_LAST = {
     'bce': 30,
-    'iou': 6,
+    'iou': 0.5,
```
Comment (joelkartoon) [inline birefnet/config.py:11]: if you are iterating / exploring these parameters, I think it's better to modify them in your bash script.

you should do this change if you have established that these are better values that we now want to use as default (maybe it's the case?)

File: birefnet/config.py [hunk 2]
```diff
@@ -76,8 +82,7 @@ def __init__(self, learning_rate=None, bce_with_logits=None, lambdas_pix_last=No
         self.log_each_steps = 15
         self.lr_warm_up_type = None
         self.display_eval_metrics = ['PA', 'BIoU', 'WF']
-
-        self.finetune_last_epochs =0
+        self.gradient_clipping_norm = 100.0
```
Comment (joelkartoon) [inline birefnet/config.py:85]: nit: make it a variable at top of script

File: scripts/evaluations.py [hunk 3]
```diff
@@ -94,27 +93,54 @@ def do_eval(args):
...
         title = prediction.split('/')
-        title.append(image_number)
+        title.append(images_number)
 
         #get the scores
         scores = get_scores(list_gt, list_pred)
 
+        #add the results to the dictionary
+        model_name=title[0]
+        weighted_scores=[images_number*s for s in scores]
```
Comment (joelkartoon) [inline scripts/evaluations.py:110]: if I understand correctly, you are computing the weighted scores for each predicted folder. It's probably easier to do it just once

File: scripts/train.py [hunk 4]
```diff
@@ -290,16 +296,18 @@ def iteration_over_batches_validation(self, epoch, info_progress=None, step_idx=
...
-                       "SSIM loss validation": loss_components_dict['ssim'],
-                       "MAE loss validation": loss_components_dict['mae'],
-                       "IoU loss validation": loss_components_dict['iou'],
-                       "GDT loss validation": loss_components_dict['gdt'],
-                       "BCE loss training": self.loss_components_train['bce'],
-                       "SSIM loss training": self.loss_components_train['ssim'],
-                       "MAE loss training": self.loss_components_train['mae'],
-                       "IoU loss training": self.loss_components_train['iou'],
-                       "GDT loss training": self.loss_components_train['gdt'],
+                       "BCE loss validation": loss_components_dict['bce'] if 'bce' in loss_components_dict else 0,
```
Comment (joelkartoon) [inline scripts/train.py:299]: `None` is a better value than `0` if the loss is absent (`0` suggests the value being 0)

File: scripts/train.py [hunk 5]
```diff
@@ -386,8 +394,10 @@ def _batch(self, batch, batch_idx, epoch, validation=False):
             if args.use_accelerate:
                 loss = loss / accelerator.gradient_accumulation_steps
                 accelerator.backward(loss)
+                accelerator.clip_grad_norm_(self.model.parameters(), config.gradient_clipping_norm)
             else:
                 loss.backward()
+                torch.nn.utils.clip_grad_norm_(self.model.parameters(), config.gradient_clipping_norm)
```
Comment (joelkartoon) [inline scripts/train.py:400]: do we need to put this line twice? can we put it outside the `if / else` loop?

File: scripts/briaai.py [hunk 6]
```diff
@@ -5,34 +5,13 @@
...
-    mask = pred_pil.resize(image.size)
-    image.putalpha(mask)
-
-    image.save(output_image_path)
+    pipe = pipeline("image-segmentation", model="briaai/RMBG-1.4", trust_remote_code=True)
+    pillow_mask = pipe(input_image_path, return_mask = True) # outputs a pillow mask
+    pillow_mask.save(output_image_path)
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser(description='Remove background from images in a folder using briaai/RMBG-2.0 model')
```
Comment (joelkartoon) [inline scripts/briaai.py:17]: nit: 1.4

File: scripts/evaluations.py [hunk 7]
```diff
@@ -128,9 +125,12 @@ def do_eval(args):
     tb.add_divider()
 
     #compute the average scores for each testset and add to the table
-    for model_name, (images_number, weighted_scores) in model_results_dict.items():
+    for model_name, model_results in model_results_dict.items():
         #compute the average scores and format them
-        average_scores=[s/images_number for s in weighted_scores]
+        images_number=sum([weight for weight, _ in model_results])
```
Comment (joelkartoon) [inline scripts/evaluations.py:130]: isolate it in a function

File: scripts/train.py [hunk 8]
```diff
@@ -394,10 +394,10 @@ def _batch(self, batch, batch_idx, epoch, validation=False):
             if args.use_accelerate:
                 loss = loss / accelerator.gradient_accumulation_steps
                 accelerator.backward(loss)
-                accelerator.clip_grad_norm_(self.model.parameters(), config.gradient_clipping_norm)
             else:
                 loss.backward()
-                torch.nn.utils.clip_grad_norm_(self.model.parameters(), config.gradient_clipping_norm)
+            clip_norm_function= accelerator.clip_grad_norm_ if args.use_accelerate else torch.nn.utils.clip_grad_norm_
```
Comment (joelkartoon) [inline scripts/train.py:399]: don't listen to me

**Review by joelkartoon — APPROVED** [review]
